### PR TITLE
Quality check workflows

### PR DESCRIPTION
**Pull request by mahesh-panchal**
Adds data and assembly quality assessment check workflows.

## PR checklist

- [ ] This comment contains a description of changes (with reason).
- [ ] If you've fixed a bug or added code that should be tested, add tests!
  - [ ] If you've added a new tool - have you followed the pipeline conventions in the [contribution docs](https://github.com/nf-core/genomeassembler/tree/master/.github/CONTRIBUTING.md)
  - [ ] If necessary, also make a PR on the nf-core/genomeassembler _branch_ on the [nf-core/test-datasets](https://github.com/nf-core/test-datasets) repository.
- [ ] Make sure your code lints (`nf-core lint`).
- [ ] Ensure the test suite passes (`nextflow run . -profile test,docker --outdir <OUTDIR>`).
- [ ] Usage Documentation in `docs/usage.md` is updated.
- [ ] Output Documentation in `docs/output.md` is updated.
- [ ] `CHANGELOG.md` is updated.
- [ ] `README.md` is updated (including new tool citations and authors/contributors).

### CODE REVIEW

**Comment by priyanka-surana** [comment]
For quality check, we do similar stuff, except an extra couple of steps that's being done here is barcoding by COI-5P and some extra cobiont checks. Another thing that's being done here about the raw HiC check is a plot of fragment separation distribution by HiLine: https://github.com/wtsi-hpag/HiLine (like the one attached). If needed all the QC checks are visible from the Tol QC page like this one: https://tolqc.cog.sanger.ac.uk/darwin/dicots/Chenopodium_album/ which makes the work more transparent and accessible.

**Comment by mahesh-panchal** [comment]
> For quality check, we do similar stuff, except an extra couple of steps that's being done here is barcoding by COI-5P and some extra cobiont checks. Another thing that's being done here about the raw HiC check is a plot of fragment separation distribution by HiLine: https://github.com/wtsi-hpag/HiLine (like the one attached). If needed all the QC checks are visible from the Tol QC page like this one: https://tolqc.cog.sanger.ac.uk/darwin/dicots/Chenopodium_album/ which makes the work more transparent and accessible.

Thanks. I'll try and incorporate those. If you have time, it would help greatly if you could contribute the module to nf-core/modules. I've seen the ToL QC portal. I have some questions about it. Are you using Mash only to compare sub libraries? Do you use KatComp, or only when you see a difference in the Mash plot? What's visualizing the samtools stats plots? For the cobiont checks, do you have links to the tools used? Also how does one interpret the Readviz classification, and contigviz classification ( I don't understand what I'm seeing there or what the density means) ?

**Comment by muffato** [comment]
Hi @mahesh-panchal . As far as I know, the code of the cobiont checks (readviz and contigviz) is not yet public. There is a brief description at https://cobiontid.github.io/ while they work on the paper.

**Comment by mahesh-panchal** [comment]
I wonder if the tests are running into disk space issues.

**Comment by mahesh-panchal** [comment]
> I wonder if the tests are running into disk space issues.

14GB available. It's not that.

**Comment by muffato** [comment]
@mahesh-panchal : Maybe you need something like https://github.com/sanger-tol/nxf-dev/pull/2/files#diff-5bc09aca193709c03f8a9752825b88bf1ca06e5d10d73da1d92d80cc1ae2fd3c ? i.e. giving read permissions for ghcr and logging in.
We needed that to use a ghcr-hosted package in GitHub actions (it was an internal package but maybe it's still needed here ?)

**Comment by mahesh-panchal** [comment]
> @mahesh-panchal : Maybe you need something like https://github.com/sanger-tol/nxf-dev/pull/2/files#diff-5bc09aca193709c03f8a9752825b88bf1ca06e5d10d73da1d92d80cc1ae2fd3c ? i.e. giving read permissions for ghcr and logging in. We needed that to use a ghcr-hosted package in GitHub actions (it was an internal package but maybe it's still needed here ?)

Thanks for the suggestion @muffato but I don't think it's that. It's using test data from the modules test data which is public, and I'm normally logged in anyway. Also the other test does pass when it's not cancelled by the edge version. In the test before last ( I think) it got through and passed.

**Review by elizabethmcd — COMMENTED** [review]
The scope of the tooling for QC looks good to me. I have one overall comment on the organization of the subworkflows integrated into the workflows. 

As is, individual pieces of the workflow are mostly written as subworkflows even if they are a single module. Whereas the original intention of subworkflows I believe is if you want to chain together multiple modules for a common task (such as samtools sort > index, etc.). So this simplifies the actual workflow script but then there are a lot of individual subworkflows which could become a lot to deal with down the road. I think how things are structured currently is fine but might be good to clarify if this is how people are expected to contribute when making PRs.

File: workflows/genomeassembler.nf [hunk 1]
```diff
@@ -68,16 +113,188 @@ workflow GENOMEASSEMBLER {
...
         ch_input
     )
-    // ch_versions = ch_versions.mix(PREPARE_INPUT.out.versions)
+    ch_versions = ch_versions.mix ( PREPARE_INPUT.out.versions )
+
+    // - TODO: Fetch Genome properties from GOAT
+
+    // BUILD KMER DATABASES
+    // builds k-mer databases as a separate step to allow reuse with resume
+    if ( params.kmer_counter == 'meryl' && ['data_qc','validate'].any { it in workflow_steps } ) {
```
[the remark this answers is not in the document]
Comment (mahesh-panchal) [reply workflows/genomeassembler.nf:135]: >  based on the input, if I provide only Hifi reads will only the database building steps be performed for hifi?
Exactly this. The PREPARE_INPUT workflow only emits tuples where there is data provided. If there is no data in a channel, the process receiving the channel does nothing. 

You can check this branch out and try it for yourself. For example, use `nextflow run main.nf -profile test_input_yml,docker -resume -dump-channels 'Input: *' -ansi-log false` to see what is in the channels, and feel free then to play with the input and see how it changes.

**Comment by mahesh-panchal** [comment]
> As is, individual pieces of the workflow are mostly written as subworkflows even if they are a single module. Whereas the original intention of subworkflows I believe is if you want to chain together multiple modules for a common task (such as samtools sort > index, etc.). So this simplifies the actual workflow script but then there are a lot of individual subworkflows which could become a lot to deal with down the road. I think how things are structured currently is fine but might be good to clarify if this is how people are expected to contribute when making PRs.

That's exactly the intention. As people contribute more modules, the subworkflows are expected to expand. However I also think it's absolutely fine to use subworkflows as a grouping and labeling mechanism, to allow for easier maintenance, and readability by describing larger units (e.g. a subworkflow is named Data properties - although this name could perhaps could do with some more thought). They are effectively channel zip-ties.

Where would you look for the information on structuring the workflow? At the moment, I'm not sure where this kind of detail would go.

**Review by d4straub — APPROVED** [review]
Looks good to me, I have only few minor comments.

**Comment by d4straub** [comment]
Hm, the test fails at random points it seems to me, I'll restart it once more.

File: workflows/genomeassembler.nf [hunk 2]
```diff
@@ -68,16 +113,188 @@ workflow GENOMEASSEMBLER {
...
+        BUILD_HIFI_MERYL_DATABASE ( PREPARE_INPUT.out.hifi )
+        BUILD_HIC_MERYL_DATABASE ( PREPARE_INPUT.out.hic)
+        BUILD_ILLUMINA_MERYL_DATABASE ( PREPARE_INPUT.out.illumina )
+        if( params.enable_ont_kmer_analyses ) {
```
[the remark this answers is not in the document]
Comment (mahesh-panchal) [reply workflows/genomeassembler.nf:139]: Yes, because ONT data is generally too noisy for k-mer analyses to apply. It's only in the current unreleased chemistry that it might be accurate enough to apply k-mer analyses. I guess one could apply read correction first and then run it through here, but I'll leave that for someone else to code if and when.

File: workflows/genomeassembler.nf [hunk 3]
```diff
@@ -68,16 +113,188 @@ workflow GENOMEASSEMBLER {
 
     ch_versions = Channel.empty()
 
+    // Check stage
```
[the remark this answers is not in the document]
Comment (mahesh-panchal) [reply workflows/genomeassembler.nf:116]: I'm reasonably sure this should be a schema check.

File: workflows/genomeassembler.nf [hunk 4]
```diff
@@ -40,8 +40,53 @@ ch_multiqc_custom_methods_description = params.multiqc_methods_description ? fil
 // SUBWORKFLOW: Consisting of a mix of local and nf-core/modules
 //
 // include { INPUT_CHECK } from '../subworkflows/local/input_check'
-include { PREPARE_INPUT } from '../subworkflows/local/prepare_input'
+include { PREPARE_INPUT } from "$projectDir/subworkflows/local/prepare_input"
```
[the remark this answers is not in the document]
Comment (mahesh-panchal) [reply workflows/genomeassembler.nf:43]: I just prefer this format of using paths relative to `$projectDir` rather than relative to current workflow. Some others prefer this too, but it's not been implemented in general because nf-core tests need to use `$launchDir` instead, and so they stuck to using relative paths. These implicit variables are not too well understood.

[the remark this answers is not in the document]
Comment (d4straub) [reply workflows/genomeassembler.nf:43]: I have no preference as long as it works.
I tested `nextflow run mahesh-panchal/genomeassembler -r quality_check_workflows -profile test,singularity --outdir results` locally and it worked. So ok to me.
